File: tldw_Server_API/app/api/v1/schemas/flashcards.py
import json
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class DeckSchedulerSettings(BaseModel):
    new_steps_minutes: list[int] = Field(default_factory=lambda: [1, 10])
    relearn_steps_minutes: list[int] = Field(default_factory=lambda: [10])
    graduating_interval_days: int = 1
    easy_interval_days: int = 4
    easy_bonus: float = 1.3
    interval_modifier: float = 1.0
    max_interval_days: int = 36500
    leech_threshold: int = 8
    enable_fuzz: bool = False


class DeckCreate(BaseModel):
    name: str = Field(..., description="Deck name (unique)")
    description: Optional[str] = Field(None, description="Deck description")
    scheduler_settings: Optional[DeckSchedulerSettings] = None


class DeckUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    scheduler_settings: Optional[DeckSchedulerSettings] = None
    expected_version: Optional[int] = Field(None, ge=1)


class Deck(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    created_at: Optional[str] = None
    last_modified: Optional[str] = None
    deleted: bool
    client_id: str
    version: int
    scheduler_settings_json: Optional[str] = None
    scheduler_settings: DeckSchedulerSettings = Field(default_factory=DeckSchedulerSettings)

    @model_validator(mode="before")
    def _populate_scheduler_settings(cls, data):
        if not isinstance(data, dict):
            return data
        if data.get("scheduler_settings") is not None:
            return data
        raw = data.get("scheduler_settings_json")
        if raw:
            try:
                parsed = json.loads(raw)
                if isinstance(parsed, dict):
                    data["scheduler_settings"] = parsed
            except Exception:
                data["scheduler_settings"] = DeckSchedulerSettings().model_dump()
        return data


class FlashcardReviewIntervalPreviews(BaseModel):
    again: str
    hard: str
    good: str
    easy: str


class FlashcardCreate(BaseModel):
    deck_id: Optional[int] = Field(None, description="Deck ID to assign the card to")
    front: str
    back: str
    notes: Optional[str] = None
    extra: Optional[str] = None
    is_cloze: Optional[bool] = False
    tags: Optional[list[str]] = Field(None, description="List of tags; stored as JSON array")
    source_ref_type: Optional[Literal['media', 'message', 'note', 'manual']] = 'manual'
    source_ref_id: Optional[str] = None
    model_type: Optional[Literal['basic','basic_reverse','cloze']] = None
    reverse: Optional[bool] = None


class Flashcard(BaseModel):
    uuid: UUID
    deck_id: Optional[int] = None
    deck_name: Optional[str] = None
    front: str
    back: str
    notes: Optional[str] = None
    extra: Optional[str] = None
    is_cloze: bool
    tags_json: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    source_ref_type: Optional[Literal['media', 'message', 'note', 'manual']] = None
    source_ref_id: Optional[str] = None
    conversation_id: Optional[str] = None
    message_id: Optional[str] = None
    ef: float
    interval_days: int
    repetitions: int
    lapses: int
    due_at: Optional[str] = None
    last_reviewed_at: Optional[str] = None
    queue_state: Literal["new", "learning", "review", "relearning", "suspended"] = "new"
    step_index: Optional[int] = None
    suspended_reason: Optional[Literal["manual", "leech"]] = None
    created_at: Optional[str] = None
    last_modified: Optional[str] = None
    deleted: bool
    client_id: str
    version: int
    model_type: Literal['basic','basic_reverse','cloze']
    reverse: bool
    next_intervals: Optional[FlashcardReviewIntervalPreviews] = None

    @model_validator(mode="before")
    def _populate_tags(cls, data):
        if not isinstance(data, dict):
            return data
        if data.get("tags") is not None:
            return data
        tags_json = data.get("tags_json")
        if tags_json:
            try:
                parsed = json.loads(tags_json)
                if isinstance(parsed, list):
                    data["tags"] = [str(t) for t in parsed if t is not None]
                else:
                    data["tags"] = []
            except Exception:
                data["tags"] = []
        else:
            data["tags"] = []
        return data


class FlashcardListResponse(BaseModel):
    items: list[Flashcard]
    count: int
    total: int | None = None


class FlashcardReviewRequest(BaseModel):
    card_uuid: str
    rating: int = Field(..., ge=0, le=5, description="Anki 0-5 rating")
    answer_time_ms: Optional[int] = None


class FlashcardReviewResponse(BaseModel):
    uuid: UUID
    ef: float
    interval_days: int
    repetitions: int
    lapses: int
    due_at: Optional[str] = None
    last_reviewed_at: Optional[str] = None
    last_modified: Optional[str] = None
    version: int
    queue_state: Literal["new", "learning", "review", "relearning", "suspended"]
    step_index: Optional[int] = None
    suspended_reason: Optional[Literal["manual", "leech"]] = None
    next_intervals: FlashcardReviewIntervalPreviews


class FlashcardNextReviewResponse(BaseModel):
    card: Optional[Flashcard] = None
    selection_reason: Optional[Literal["learning_due", "review_due", "new", "none"]] = None


class FlashcardGenerateRequest(BaseModel):
    text: str = Field(..., min_length=1, description="Source text to generate flashcards from")
    num_cards: int = Field(10, ge=1, le=100, description="Requested number of generated cards")
    card_type: Literal['basic', 'basic_reverse', 'cloze'] = Field('basic')
    difficulty: Literal['easy', 'medium', 'hard', 'mixed'] = Field('mixed')
    focus_topics: list[str] = Field(default_factory=list)
    provider: Optional[str] = Field(None, description="Optional LLM provider override")
    model: Optional[str] = Field(None, description="Optional LLM model override")


class GeneratedFlashcard(BaseModel):
    front: str
    back: str
    tags: list[str] = Field(default_factory=list)
    model_type: Literal['basic', 'basic_reverse', 'cloze'] = Field('basic')
    notes: Optional[str] = None
    extra: Optional[str] = None


class FlashcardGenerateResponse(BaseModel):
    flashcards: list[GeneratedFlashcard] = Field(default_factory=list)
    count: int


class FlashcardDeckProgress(BaseModel):
    deck_id: int
    deck_name: str
    total: int
    new: int
    learning: int
    due: int
    mature: int


class FlashcardAnalyticsSummaryResponse(BaseModel):
    reviewed_today: int
    retention_rate_today: Optional[float] = None
    lapse_rate_today: Optional[float] = None
    avg_answer_time_ms_today: Optional[float] = None
    study_streak_days: int
    generated_at: str
    decks: list[FlashcardDeckProgress] = Field(default_factory=list)


class FlashcardUpdate(BaseModel):
    deck_id: Optional[int] = None
    front: Optional[str] = None
    back: Optional[str] = None
    notes: Optional[str] = None
    extra: Optional[str] = None
    is_cloze: Optional[bool] = None
    tags: Optional[list[str]] = None  # Optional: set/replace tags
    expected_version: Optional[int] = None
    model_type: Optional[Literal['basic','basic_reverse','cloze']] = None
    reverse: Optional[bool] = None


class FlashcardBulkUpdateItem(FlashcardUpdate):
    uuid: str


class FlashcardBulkUpdateError(BaseModel):
    code: Literal["validation_error", "not_found", "conflict"]
    message: str
    invalid_fields: list[str] = Field(default_factory=list)
    invalid_deck_ids: list[int] = Field(default_factory=list)


class FlashcardBulkUpdateResult(BaseModel):
    uuid: str
    status: Literal["updated", "validation_error", "not_found", "conflict"]
    flashcard: Optional[Flashcard] = None
    error: Optional[FlashcardBulkUpdateError] = None


class FlashcardBulkUpdateResponse(BaseModel):
    results: list[FlashcardBulkUpdateResult] = Field(default_factory=list)


class FlashcardAssetMetadata(BaseModel):
    asset_uuid: UUID
    reference: str
    markdown_snippet: str
    mime_type: str
    byte_size: int
    width: Optional[int] = None
    height: Optional[int] = None
    original_filename: Optional[str] = None


class FlashcardResetSchedulingRequest(BaseModel):
    expected_version: int = Field(..., ge=1)


class StudyAssistantThreadSummary(BaseModel):
    id: int
    context_type: Literal["flashcard", "quiz_attempt_question"]
    flashcard_uuid: Optional[str] = None
    quiz_attempt_id: Optional[int] = None
    question_id: Optional[int] = None
    last_message_at: Optional[str] = None
    message_count: int = 0
    deleted: bool
    client_id: str
    version: int
    created_at: Optional[str] = None
    last_modified: Optional[str] = None


class StudyAssistantMessage(BaseModel):
    id: int
    thread_id: int
    role: Literal["user", "assistant"]
    action_type: Literal["explain", "mnemonic", "follow_up", "fact_check", "freeform"]
    input_modality: Literal["text", "voice_transcript"]
    content: str
    structured_payload: dict[str, Any] = Field(default_factory=dict)
    context_snapshot: dict[str, Any] = Field(default_factory=dict)
    provider: Optional[str] = None
    model: Optional[str] = None
    created_at: Optional[str] = None
    client_id: str

    @model_validator(mode="before")
    def _populate_json_fields(cls, data):
        if not isinstance(data, dict):
            return data

        for source_field, target_field in (
            ("structured_payload_json", "structured_payload"),
            ("context_snapshot_json", "context_snapshot"),
        ):
            if data.get(target_field) is not None:
                continue
            raw = data.get(source_field)
            if isinstance(raw, dict):
                data[target_field] = raw
                continue
            if isinstance(raw, str):
                try:
                    parsed = json.loads(raw)
                except Exception:
                    parsed = {}
                data[target_field] = parsed if isinstance(parsed, dict) else {}

        if data.get("structured_payload") is None:
            data["structured_payload"] = {}
        if data.get("context_snapshot") is None:
            data["context_snapshot"] = {}
        return data


class StudyAssistantHistoryResponse(BaseModel):
    thread: StudyAssistantThreadSummary
    messages: list[StudyAssistantMessage] = Field(default_factory=list)


StudyAssistantAction = Literal["explain", "mnemonic", "follow_up", "fact_check", "freeform"]


class StudyAssistantFactCheckPayload(BaseModel):
    verdict: Literal["correct", "partially_correct", "incorrect"]
    corrections: list[str] = Field(default_factory=list)
    missing_points: list[str] = Field(default_factory=list)
    next_prompt: str = Field(default="What part would you like to review next?")


class StudyAssistantRespondRequest(BaseModel):
    action: StudyAssistantAction
    message: Optional[str] = None
    input_modality: Literal["text", "voice_transcript"] = "text"
    provider: Optional[str] = None
    model: Optional[str] = None
    expected_thread_version: Optional[int] = Field(None, ge=1)


class StudyAssistantContextResponse(BaseModel):
    thread: StudyAssistantThreadSummary
    messages: list[StudyAssistantMessage] = Field(default_factory=list)
    context_snapshot: dict[str, Any] = Field(default_factory=dict)
    available_actions: list[StudyAssistantAction] = Field(default_factory=list)


class StudyAssistantRespondResponse(BaseModel):
    thread: StudyAssistantThreadSummary
    user_message: StudyAssistantMessage
    assistant_message: StudyAssistantMessage
    structured_payload: dict[str, Any] = Field(default_factory=dict)
    context_snapshot: dict[str, Any] = Field(default_factory=dict)


class FlashcardTagsUpdate(BaseModel):
    tags: list[str]


class FlashcardQuery(BaseModel):
    deck_id: Optional[int] = None
    tag: Optional[str] = None
    due_status: Optional[Literal['new', 'learning', 'due', 'all']] = 'all'
    q: Optional[str] = None
    limit: Optional[int] = Field(100, ge=1, le=1000)
    offset: Optional[int] = Field(0, ge=0)
    order_by: Optional[Literal['due_at', 'created_at']] = 'due_at'


class FlashcardsImportRequest(BaseModel):
    content: str = Field(..., description="TSV content: Deck, Front, Back, Tags, Notes per line")
    delimiter: Optional[str] = Field('\t', description="Field delimiter; default tab")
    has_header: Optional[bool] = Field(False, description="Whether the first line is a header")


class StructuredQaImportPreviewRequest(BaseModel):
    content: str = Field(..., min_length=1, description="Labeled Q&A content for preview parsing")


class StructuredQaImportPreviewDraft(BaseModel):
    front: str
    back: str
    line_start: int
    line_end: int
    notes: Optional[str] = None
    extra: Optional[str] = None
    tags: list[str] = Field(default_factory=list)


class StructuredQaImportPreviewError(BaseModel):
    line: Optional[int] = None
    error: str


class StructuredQaImportPreviewResponse(BaseModel):
    drafts: list[StructuredQaImportPreviewDraft] = Field(default_factory=list)
    errors: list[StructuredQaImportPreviewError] = Field(default_factory=list)
    detected_format: Literal["qa_labels"] = "qa_labels"
    skipped_blocks: int = 0

File: apps/packages/ui/src/services/flashcards.ts
import { bgRequest, bgUpload } from "@/services/background-proxy"
import type { AllowedPath } from "@/services/tldw/openapi-guard"
import {
  buildQuery,
  createResourceClient
} from "@/services/resource-client"

const decksClient = createResourceClient({
  basePath: "/api/v1/flashcards/decks" as AllowedPath
})

const flashcardsClient = createResourceClient({
  basePath: "/api/v1/flashcards" as AllowedPath
})

export type DeckSchedulerSettings = {
  new_steps_minutes: number[]
  relearn_steps_minutes: number[]
  graduating_interval_days: number
  easy_interval_days: number
  easy_bonus: number
  interval_modifier: number
  max_interval_days: number
  leech_threshold: number
  enable_fuzz: boolean
}

export type DeckSchedulerType = "sm2_plus" | "fsrs"

export type FsrsSchedulerSettings = {
  target_retention: number
  maximum_interval_days: number
  enable_fuzz: boolean
}

export type DeckSchedulerSettingsEnvelope = {
  sm2_plus: DeckSchedulerSettings
  fsrs: FsrsSchedulerSettings
}

export type DeckSchedulerSettingsEnvelopeUpdate = {
  sm2_plus?: Partial<DeckSchedulerSettings>
  fsrs?: Partial<FsrsSchedulerSettings>
}

export type FlashcardIntervalPreviews = {
  again: string
  hard: string
  good: string
  easy: string
}

export type StudyAssistantAction =
  | "explain"
  | "mnemonic"
  | "follow_up"
  | "fact_check"
  | "freeform"

export type StudyAssistantInputModality = "text" | "voice_transcript"

export type StudyAssistantThreadSummary = {
  id: number
  context_type: "flashcard" | "quiz_attempt_question"
  flashcard_uuid?: string | null
  quiz_attempt_id?: number | null
  question_id?: number | null
  last_message_at?: string | null
  message_count: number
  deleted: boolean
  client_id: string
  version: number
  created_at?: string | null
  last_modified?: string | null
}

export type StudyAssistantMessage = {
  id: number
  thread_id: number
  role: "user" | "assistant"
  action_type: StudyAssistantAction
  input_modality: StudyAssistantInputModality
  content: string
  structured_payload: Record<string, unknown>
  context_snapshot: Record<string, unknown>
  provider?: string | null
  model?: string | null
  created_at?: string | null
  client_id: string
}

export type StudyAssistantFactCheckPayload = {
  verdict: "correct" | "partially_correct" | "incorrect"
  corrections: string[]
  missing_points: string[]
  next_prompt: string
}

export type StudyAssistantRespondRequest = {
  action: StudyAssistantAction
  message?: string | null
  input_modality?: StudyAssistantInputModality
  provider?: string | null
  model?: string | null
  expected_thread_version?: number | null
}

export type StudyAssistantContextResponse = {
  thread: StudyAssistantThreadSummary
  messages: StudyAssistantMessage[]
  context_snapshot: Record<string, unknown>
  available_actions: StudyAssistantAction[]
}

export type StudyAssistantRespondResponse = {
  thread: StudyAssistantThreadSummary
  user_message: StudyAssistantMessage
  assistant_message: StudyAssistantMessage
  structured_payload: Record<string, unknown>
  context_snapshot: Record<string, unknown>
}

// Minimal client types based on openapi.json
export type Deck = {
  id: number
  name: string
  description?: string | null
  deleted: boolean
  client_id: string
  version: number
  created_at?: string | null
  last_modified?: string | null
  scheduler_type: DeckSchedulerType
  scheduler_settings_json?: string | null
  scheduler_settings: DeckSchedulerSettingsEnvelope
}

export type Flashcard = {
  uuid: string
  deck_id?: number | null
  front: string
  back: string
  notes?: string | null
  extra?: string | null
  is_cloze: boolean
  tags?: string[] | null
  ef: number
  interval_days: number
  repetitions: number
  lapses: number
  due_at?: string | null
  created_at?: string | null
  last_reviewed_at?: string | null
  queue_state: "new" | "learning" | "review" | "relearning" | "suspended"
  step_index?: number | null
  suspended_reason?: "manual" | "leech" | null
  last_modified?: string | null
  deleted: boolean
  client_id: string
  version: number
  model_type: "basic" | "basic_reverse" | "cloze"
  reverse: boolean
  scheduler_type?: DeckSchedulerType | null
  source_ref_type?: "media" | "message" | "note" | "manual" | null
  source_ref_id?: string | null
  conversation_id?: string | null
  message_id?: string | null
  next_intervals?: FlashcardIntervalPreviews | null
}

export type DeckUpdate = {
  name?: string | null
  description?: string | null
  scheduler_type?: DeckSchedulerType | null
  scheduler_settings?: DeckSchedulerSettingsEnvelopeUpdate | null
  expected_version?: number | null
}

export type DeckCreateInput = {
  name: string
  description?: string | null
  scheduler_type?: DeckSchedulerType | null
  scheduler_settings?: DeckSchedulerSettingsEnvelope | null
}

export type FlashcardCreate = {
  deck_id?: number | null
  front: string
  back: string
  notes?: string | null
  extra?: string | null
  is_cloze?: boolean | null
  tags?: string[] | null
  source_ref_type?: "media" | "message" | "note" | "manual" | null
  source_ref_id?: string | null
  model_type?: Flashcard["model_type"] | null
  reverse?: boolean | null
}

export type FlashcardUpdate = {
  deck_id?: number | null
  front?: string | null
  back?: string | null
  notes?: string | null
  extra?: string | null
  is_cloze?: boolean | null
  tags?: string[] | null
  expected_version?: number | null
  model_type?: Flashcard["model_type"] | null
  reverse?: boolean | null
}

export type FlashcardBulkUpdateItem = FlashcardUpdate & {
  uuid: string
}

export type FlashcardBulkUpdateError = {
  code: "validation_error" | "not_found" | "conflict"
  message: string
  invalid_fields?: string[]
  invalid_deck_ids?: number[]
}

export type FlashcardBulkUpdateResult = {
  uuid: string
  status: "updated" | "validation_error" | "not_found" | "conflict"
  flashcard?: Flashcard | null
  error?: FlashcardBulkUpdateError | null
}

export type FlashcardBulkUpdateResponse = {
  results: FlashcardBulkUpdateResult[]
}

export type FlashcardResetSchedulingRequest = {
  expected_version: number
}

export type FlashcardListResponse = {
  items: Flashcard[]
  count: number
  total?: number | null
}

export type FlashcardReviewRequest = {
  card_uuid: string
  rating: number // 0-5
  answer_time_ms?: number | null
}

export type FlashcardGeneratedDraft = {
  front: string
  back: string
  tags?: string[] | null
  model_type?: "basic" | "basic_reverse" | "cloze"
  notes?: string | null
  extra?: string | null
}

export type FlashcardsGenerateRequest = {
  text: string
  num_cards?: number
  card_type?: "basic" | "basic_reverse" | "cloze"
  difficulty?: "easy" | "medium" | "hard" | "mixed"
  focus_topics?: string[] | null
  provider?: string | null
  model?: string | null
}

export type FlashcardsGenerateResponse = {
  flashcards: FlashcardGeneratedDraft[]
  count: number
}

export type FlashcardReviewResponse = {
  uuid: string
  ef: number
  interval_days: number
  repetitions: number
  lapses: number
  due_at?: string | null
  last_reviewed_at?: string | null
  last_modified?: string | null
  version: number
  scheduler_type: DeckSchedulerType
  queue_state: Flashcard["queue_state"]
  step_index?: number | null
  suspended_reason?: Flashcard["suspended_reason"]
  next_intervals: FlashcardIntervalPreviews
}

export type FlashcardNextReviewResponse = {
  card?: Flashcard | null
  selection_reason?: "learning_due" | "review_due" | "new" | "none" | null
}

export type FlashcardsImportRequest = {
  content: string
  delimiter?: string | null
  has_header?: boolean | null
}

export type FlashcardsImportJsonRequest = {
  content: string
  filename?: string | null
}

export type FlashcardsImportApkgRequest = {
  bytes: Uint8Array
  filename?: string | null
}

export type StructuredQaImportPreviewRequest = {
  content: string
}

export type FlashcardsImportError = {
  line?: number | null
  index?: number | null
  error: string
}

export type FlashcardsImportResponse = {
  imported: number
  items: Array<{
    uuid: string
    deck_id: number
  }>
  errors: FlashcardsImportError[]
}

export type StructuredQaImportPreviewDraft = {
  front: string
  back: string
  line_start: number
  line_end: number
  notes?: string | null
  extra?: string | null
  tags?: string[] | null
}

export type StructuredQaImportPreviewResponse = {
  drafts: StructuredQaImportPreviewDraft[]
  errors: Array<{ line?: number | null; error: string }>
  detected_format: "qa_labels"
  skipped_blocks: number
}

export type FlashcardsExportParams = {
  deck_id?: number | null
  tag?: string | null
  q?: string | null
  format?: "csv" | "apkg" | null
  include_reverse?: boolean | null
  delimiter?: string | null
  include_header?: boolean | null
  extended_header?: boolean | null
}

export type FlashcardDeckProgress = {
  deck_id: number
  deck_name: string
  total: number
  new: number
  learning: number
  due: number
  mature: number
}

export type FlashcardAnalyticsSummary = {
  reviewed_today: number
  retention_rate_today?: number | null
  lapse_rate_today?: number | null
  avg_answer_time_ms_today?: number | null
  study_streak_days: number
  generated_at: string
  decks: FlashcardDeckProgress[]
}

// Decks
export async function listDecks(options?: { signal?: AbortSignal }): Promise<Deck[]> {
  return await decksClient.list<Deck[]>(undefined, {
    abortSignal: options?.signal
  })
}

export async function createDeck(
  input: DeckCreateInput,
  options?: { signal?: AbortSignal }
): Promise<Deck> {
  return await decksClient.create<Deck>(input, {
    abortSignal: options?.signal
  })
}

export async function updateDeck(
  deck_id: number,
  input: DeckUpdate,
  options?: { signal?: AbortSignal }
): Promise<Deck> {
  return await decksClient.update<Deck>(String(deck_id), input, {
    abortSignal: options?.signal
  })
}

// Flashcards CRUD
export async function listFlashcards(params: {
  deck_id?: number | null
  tag?: string | null
  due_status?: "new" | "learning" | "due" | "all" | null
  q?: string | null
  limit?: number
  offset?: number
  order_by?: "due_at" | "created_at" | null
}): Promise<FlashcardListResponse> {
  return await flashcardsClient.list<FlashcardListResponse>({
    deck_id: params.deck_id,
    tag: params.tag,
    due_status: params.due_status,
    q: params.q,
    limit: params.limit,
    offset: params.offset,
    order_by: params.order_by
  })
}

export async function createFlashcard(
  input: FlashcardCreate,
  options?: { signal?: AbortSignal }
): Promise<Flashcard> {
  return await flashcardsClient.create<Flashcard>(input, {
    abortSignal: options?.signal
  })
}

export async function createFlashcardsBulk(
  input: FlashcardCreate[]
): Promise<FlashcardListResponse> {
  return await bgRequest<FlashcardListResponse, AllowedPath, "POST">({
    path: "/api/v1/flashcards/bulk",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input
  })
}

export async function updateFlashcardsBulk(
  input: FlashcardBulkUpdateItem[]
): Promise<FlashcardBulkUpdateResponse> {
  return await bgRequest<FlashcardBulkUpdateResponse, AllowedPath, "PATCH">({
    path: "/api/v1/flashcards/bulk",
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: input
  })
}

export async function getFlashcard(card_uuid: string): Promise<Flashcard> {
  return await flashcardsClient.get<Flashcard>(card_uuid)
}

export async function getFlashcardAssistant(
  card_uuid: string,
  options?: { signal?: AbortSignal }
): Promise<StudyAssistantContextResponse> {
  return await bgRequest<StudyAssistantContextResponse, AllowedPath, "GET">({
    path: `/api/v1/flashcards/${card_uuid}/assistant` as AllowedPath,
    method: "GET",
    abortSignal: options?.signal
  })
}

export async function respondFlashcardAssistant(
  card_uuid: string,
  input: StudyAssistantRespondRequest,
  options?: { signal?: AbortSignal }
): Promise<StudyAssistantRespondResponse> {
  return await bgRequest<StudyAssistantRespondResponse, AllowedPath, "POST">({
    path: `/api/v1/flashcards/${card_uuid}/assistant/respond` as AllowedPath,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input,
    abortSignal: options?.signal
  })
}

export async function updateFlashcard(card_uuid: string, input: FlashcardUpdate): Promise<void> {
  await flashcardsClient.update<void>(card_uuid, input)
}

export async function deleteFlashcard(card_uuid: string, expected_version: number): Promise<void> {
  await flashcardsClient.remove<void>(card_uuid, {
    expected_version
  })
}

export async function resetFlashcardScheduling(
  card_uuid: string,
  input: FlashcardResetSchedulingRequest
): Promise<Flashcard> {
  return await bgRequest<Flashcard, AllowedPath, "POST">({
    path: `/api/v1/flashcards/${card_uuid}/reset-scheduling` as AllowedPath,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input
  })
}

// Review
export async function reviewFlashcard(input: FlashcardReviewRequest): Promise<FlashcardReviewResponse> {
  return await bgRequest<FlashcardReviewResponse, AllowedPath, "POST">({
    path: "/api/v1/flashcards/review",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input
  })
}

export async function getNextReviewCard(
  deck_id?: number | null
): Promise<FlashcardNextReviewResponse> {
  const query = buildQuery({
    deck_id
  })
  return await bgRequest<FlashcardNextReviewResponse, AllowedPath, "GET">({
    path: `/api/v1/flashcards/review/next${query}` as AllowedPath,
    method: "GET"
  })
}

export async function generateFlashcards(
  input: FlashcardsGenerateRequest
): Promise<FlashcardsGenerateResponse> {
  return await bgRequest<FlashcardsGenerateResponse, AllowedPath, "POST">({
    path: "/api/v1/flashcards/generate",
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: input
  })
}

// Import
export async function getFlashcardsImportLimits(): Promise<any> {
  return await bgRequest<any, AllowedPath, "GET">({
    path: "/api/v1/config/flashcards-import-limits",
    method: "GET"
  })
}

export async function importFlashcards(payload: FlashcardsImportRequest, overrides?: {
  max_lines?: number | null
  max_line_length?: number | null
  max_field_length?: number | null
}): Promise<FlashcardsImportResponse> {
  const query = buildQuery({
    max_lines: overrides?.max_lines,
    max_line_length: overrides?.max_line_length,
    max_field_length: overrides?.max_field_length
  })
  const path = `/api/v1/flashcards/import${query}` as AllowedPath
  return await bgRequest<FlashcardsImportResponse, AllowedPath, "POST">({
    path,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: payload
  })
}

export async function previewStructuredQaImport(
  payload: StructuredQaImportPreviewRequest,
  overrides?: {
    max_lines?: number | null
    max_line_length?: number | null
    max_field_length?: number | null
  }
): Promise<StructuredQaImportPreviewResponse> {
  const query = buildQuery({
    max_lines: overrides?.max_lines,
    max_line_length: overrides?.max_line_length,
    max_field_length: overrides?.max_field_length
  })
  const path = `/api/v1/flashcards/import/structured/preview${query}` as AllowedPath
  return await bgRequest<StructuredQaImportPreviewResponse, AllowedPath, "POST">({
    path,
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: payload
  })
}

export async function importFlashcardsJson(
  payload: FlashcardsImportJsonRequest,
  overrides?: {
    max_items?: number | null
    max_field_length?: number | null
  }
): Promise<FlashcardsImportResponse> {
  const query = buildQuery({
    max_items: overrides?.max_items,
    max_field_length: overrides?.max_field_length
  })
  const path = `/api/v1/flashcards/import/json${query}` as AllowedPath
  const filename = (payload.filename || "flashcards.json").trim() || "flashcards.json"
  const lowerName = filename.toLowerCase()
  const mimeType =
    lowerName.endsWith(".jsonl") || lowerName.endsWith(".ndjson")
      ? "application/x-ndjson"
      : "application/json"
  const bytes = new TextEncoder().encode(payload.content)

  return await bgUpload<FlashcardsImportResponse, AllowedPath, "POST">({
    path,
    method: "POST",
    fileFieldName: "file",
    file: {
      name: filename,
      type: mimeType,
      data: bytes
    }
  })
}

export async function importFlashcardsApkg(
  payload: FlashcardsImportApkgRequest,
  overrides?: {
    max_items?: number | null
    max_field_length?: number | null
  }
): Promise<FlashcardsImportResponse> {
  const query = buildQuery({
    max_items: overrides?.max_items,
    max_field_length: overrides?.max_field_length
  })
  const path = `/api/v1/flashcards/import/apkg${query}` as AllowedPath
  const filename = (payload.filename || "flashcards.apkg").trim() || "flashcards.apkg"

  return await bgUpload<FlashcardsImportResponse, AllowedPath, "POST">({
    path,
    method: "POST",
    fileFieldName: "file",
    file: {
      name: filename,
      type: "application/apkg",
      data: payload.bytes
    }
  })
}

export async function getFlashcardsAnalyticsSummary(params?: {
  deck_id?: number | null
  signal?: AbortSignal
}): Promise<FlashcardAnalyticsSummary> {
  const query = buildQuery({
    deck_id: params?.deck_id
  })
  const path = `/api/v1/flashcards/analytics/summary${query}` as AllowedPath
  return await bgRequest<FlashcardAnalyticsSummary, AllowedPath, "GET">({
    path,
    method: "GET",
    abortSignal: params?.signal
  })
}

// Export (returns text/csv or file-like payload)
export async function exportFlashcards(params: FlashcardsExportParams = {}): Promise<string> {
  const query = buildQuery({
    deck_id: params.deck_id,
    tag: params.tag,
    q: params.q,
    format: params.format,
    include_reverse: params.include_reverse,
    delimiter: params.delimiter,
    include_header: params.include_header,
    extended_header: params.extended_header
  })
  const path = `/api/v1/flashcards/export${query}` as AllowedPath
  // Force accept text so bgRequest returns text
  return await bgRequest<string, AllowedPath, "GET">({
    path,
    method: "GET",
    headers: { Accept: "text/plain, text/csv, application/octet-stream, application/json;q=0.5" }
  })
}

// Export binary (APKG). Uses direct fetch to preserve binary payload.
export async function exportFlashcardsFile(params: FlashcardsExportParams & { format: 'apkg' }): Promise<Blob> {
  const query = buildQuery({
    deck_id: params.deck_id,
    tag: params.tag,
    q: params.q,
    format: "apkg",
    include_reverse: params.include_reverse,
    // CSV specific options ignored for apkg on server side, but safe to pass
    delimiter: params.delimiter,
    include_header: params.include_header,
    extended_header: params.extended_header
  })
  const path = `/api/v1/flashcards/export${query}` as AllowedPath
  const response = await bgRequest<{
    ok: boolean
    status: number
    data?: ArrayBuffer
    error?: string
    headers?: Record<string, string>
  }>({
    path,
    method: "GET",
    headers: { Accept: "application/octet-stream" },
    responseType: "arrayBuffer",
    returnResponse: true
  })
  if (!response) {
    throw new Error("Export failed")
  }
  if (!response.ok) {
    throw new Error(response.error || `Export failed: ${response.status}`)
  }
  const headers = new Headers(response.headers || {})
  return new Blob([response.data ?? new Uint8Array()], {
    type: headers.get("content-type") || "application/octet-stream"
  })
}
